Simple `SIGQUIT` support on Windows

staticfloat opened the pull request:
Attempting to call `MiniDumpWriteDump()` to get crash dumps on windows when SIGQUIT is passed.

DilumAluthge commented:
Would it be worth upstreaming this?

vtjnash commented on src/win/process.c:
```diff
@@ -1251,6 +1251,18 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        NULL,
+        NULL,
+        NULL,
+      );
```
I am curious if this would work better (`MiniDumpWriteDump` claims to be single threaded):
```
HREPORT report;
WerReportCreate(L"SIGQUIT", WerReportNonCritical, NULL, &report)
WerReportAddDump(report, process_handle, NULL, WerDumpTypeMiniDump, NULL, 0);
WerReportSubmit(report, WerConsentDenied, 0, 0, NULL);
WerReportCloseHandle(report);
```
The user then may be able to enable it manually via:
https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps?source=recommendations

Note: completely untested code, just written by reading the docs and hoping they are accurate.

staticfloat replied:
Just to close the loop publicly, we spent quite a while trying to coerce WER to output dumps locally, but it just didn't work well at all.  We have decided instead to use `MiniDumpWriteDump()` in combination with a little bit of windows registry parsing to emulate similar behavior, so as to get local dumps written out to a stable location for both uncontrolled exits, as well as `SIGQUIT`'s.

vtjnash commented on src/win/process.c:
```diff
@@ -1251,6 +1251,18 @@ static int uv__kill(HANDLE process_handle, int signum) {
+      MiniDumpWriteDump(
+        process_handle,
+        GetProcessId(process_handle),
+        MiniDumpNormal,
```
```suggestion
        "need a file path here.minidump",
        MiniDumpNormal,
```

staticfloat commented:
I could not, for the life of me, get WER to generate a dump file.  I could generate `.wer` files, and they claimed to have heap dumps attached, but I could never find where they were.  I have reverted to using `MiniDumpWriteDump()`, which is confirmed working.  The next question is; do we want to provide a configuration mechanism?  I think I have chosen sane defaults for most things, but perhaps we want to be able to customize the directory the dumps are written out to, the kind of dump that is generated, the filename of the dump, etc...?

staticfloat commented:
For the record, the way I confirmed this is working was to build Julia with this patched libuv, then run:

```julia
p = run(`$(Base.julia_cmd()) -e "sleep(10)"`; wait=false)
sleep(1)
kill(p, Base.SIGQUIT)
```

The generated `.dmp` file is then openable in `windbg`.

One downside to not using WER is that we don't get a unified dump collection mechanism for both `SIGQUIT`-initiated dumps as well as segfault-initiated dumps.  I'd really like to be able to do both, but I just haven't been able to get them to work.  :(

vtjnash commented on configure.ac:
```diff
@@ -74,7 +74,7 @@ AM_CONDITIONAL([OS400],    [AS_CASE([$host_os],[os400],         [true], [false])
 AM_CONDITIONAL([WINNT],    [AS_CASE([$host_os],[mingw*],        [true], [false])])
 AS_CASE([$host_os],[mingw*], [
-    LIBS="$LIBS -lws2_32 -lpsapi -liphlpapi -lshell32 -lsecur32 -luserenv -luser32"
+    LIBS="$LIBS -lws2_32 -lpsapi -liphlpapi -lshell32 -lsecur32 -luserenv -luser32 -lwer -ldbghelp"
```
```suggestion
    LIBS="$LIBS -lws2_32 -lpsapi -liphlpapi -lshell32 -lsecur32 -luserenv -luser32 -ldbghelp -lole32 -luuid"
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1252,68 @@ static int uv__kill(HANDLE process_handle, int signum) {
     return UV_EINVAL;
   }
 
+  // Create a dump file for the targeted process, if the registry key
```
use C-style comments only (/**/), and full sentences (end with a period)

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1252,68 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        CoTaskMemFree(localappdata);
+      }
+      RegCloseKey(registry_key);
+    }
```
```suggestion
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1252,68 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        // Default value for `dump_folder` is `%LOCALAPPDATA%\CrashDumps`
+        WCHAR * localappdata;
+        SHGetKnownFolderPath(&FOLDERID_LocalAppData, 0, NULL, &localappdata);
+        swprintf_s(dump_folder, sizeof(dump_folder), L"%ls\\CrashDumps", localappdata);
```
check return values of `swprintf_s` to make sure it fit?

vtjnash replied:
Also, call it `_snwprintf`

staticfloat replied:
I don't think checking the return value is useful, since if it didn't fit, we will just fail the `CreateFile` later on, right?

vtjnash replied:
oh, that might require calling `_snwprintf_s` then?

vtjnash replied:
there are far too many ways to call this function on windows, and I don't know which one is right, other than attempting to see what was used before

vtjnash replied:
once upon a time...

commit c6ecf97aafc858c2ad1089fb78da6c586d61d8b6
Author: Ben Noordhuis <info@bnoordhuis.nl>
Date:   Fri Nov 8 12:22:37 2013 +0100

    windows: use _snwprintf(), not swprintf()
    
    Drop the _CRT_NON_CONFORMING_SWPRINTFS hack and just use _snwprintf().
    
    It's a long and complicated story but the gist of it is that the MS CRT
    had a swprintf() function before ISO C did, with a different function
    prototype to boot: the ISO C one takes a |size| argument, the MS one
    does not.
    
    The function prototype that's exported by mingw and mingw-w64 depends
    on the mingw version and the _CRT_NON_CONFORMING_SWPRINTFS define.
    If they don't match up, you get the wrong prototype and things will
    crash at run-time.
    
    Reduce the phase space by sidestepping the whole issue: drop swprintf()
    altogether and use _snwprintf() from now on.
    
    Fixes #990.

staticfloat commented on src/win/process.c:
```diff
@@ -1250,7 +1252,68 @@ static int uv__kill(HANDLE process_handle, int signum) {
+  // `HKLM:Software\Microsoft\Windows\Windows Error Reporting\LocalDumps\DumpFolder`
+  // points to a valid folder.  The default value if the registry key does not
+  // exist is `%LOCALAPPDATA%\CrashDumps`, see [0] for more detail.
+  // [0]: https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps
```
```suggestion
  /*
   * Create a dump file for the targeted process, if the registry key
   * `HKLM:Software\Microsoft\Windows\Windows Error Reporting\LocalDumps\DumpFolder`
   * points to a valid folder.  The default value if the registry key does not
   * exist is `%LOCALAPPDATA%\CrashDumps`, see [0] for more detail.
   * [0]: https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps
   */
```

staticfloat commented on src/win/process.c:
```diff
@@ -1250,7 +1252,68 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        // Default value for `dump_folder` is `%LOCALAPPDATA%\CrashDumps`
+        WCHAR * localappdata;
+        SHGetKnownFolderPath(&FOLDERID_LocalAppData, 0, NULL, &localappdata);
+        swprintf_s(dump_folder, sizeof(dump_folder), L"%ls\\CrashDumps", localappdata);
```
```suggestion
        _snwprintf(dump_folder, sizeof(dump_folder), L"%ls\\CrashDumps", localappdata);
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1252,69 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        process_handle, pid,
+        hDumpFile, MiniDumpWithFullMemory | MiniDumpIgnoreInaccessibleMemory | MiniDumpWithAvxXStateContext,
+        NULL, NULL, NULL
+      );
```
```suggestion
      if (MiniDumpWriteDump(
        process_handle, pid,
        hDumpFile, MiniDumpWithFullMemory | MiniDumpIgnoreInaccessibleMemory | MiniDumpWithAvxXStateContext,
        NULL, NULL, NULL
      )) {
        FILE_DISPOSITION_INFO DeleteOnClose = { FALSE };
        SetFileInformationByHandle(hDumpFile, FileDispositionInfo, &DeleteOnClose, sizeof(DeleteOnClose);
      }
```
(and also open the file with `FILE_FLAG_DELETE_ON_CLOSE`)

vtjnash reviewed (commented):
Can you submit this upstream to v1.x also?

staticfloat committed:
[Windows] Write out minidumps on reception of SIGQUIT

This commit adds the ability to dump core when sending the `SIGQUIT`
signal on Windows.  The change reads in the current registry setting for
local dumps, and attempts to write out to that location before killing
the process.  See [0] for registry and pathing details. This behavior
mimics that of the dumps created by the typical Windows
Error Reporting mechanism.

[0] https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps

staticfloat commented on src/win/process.c:
```diff
@@ -1250,7 +1253,82 @@ static int uv__kill(HANDLE process_handle, int signum) {
+    WCHAR dump_name[MAX_PATH];
+    _snwprintf_s(dump_name, sizeof(dump_name), _TRUNCATE, L"%ls\\%ls.%d.dmp", dump_folder, basename, pid);
+
+    HANDLE hDumpFile = CreateFileW(dump_name, GENERIC_WRITE, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
```
@vtjnash I was doing some more strenuous testing of this, and I discovered that WER actually is a little more aggressive than us; it will try to create the `dump_folder` directory if it does not exist.

Also, the gate for WER to create a local dump at all is the presence of the `LocalDumps` registry key.

My proposed behavior change here is to:

1) Check that `HKLM:SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps` exists at all, and if it does not, break.
2) Try to create the crash dumps folder if it doesn't already exist.

vtjnash replied:
I think I am okay with deviating slightly there, if useful. But (1) is definitely needed, since it appears you are accessing dump_folder undefined here

staticfloat replied:
Alright, I figured out how to properly emulate WER here, and it's quite simple.  I think we've got full compatibility now, as far as this goes.  We're not emulating the process-name-specific overrides, but I'm okay with leaving that for future work.  :)

staticfloat commented:
@vtjnash could you review once more, and if it's good, I'll submit to upstream on the `v1.x` branch.

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+    DWORD pid = GetProcessId(process_handle);
+
+    /* Get LocalDumps directory path. */
+    HKEY registry_key;
```
style nit: move this decl to the top of the scope

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+    GetModuleBaseNameW(process_handle, NULL, &basename[0], sizeof(basename));
+
+    /* Get PID of target process. */
+    DWORD pid = GetProcessId(process_handle);
```
style nit: don't mix code and decl on the same line (but also keep all decls together)
```suggestion
    DWORD pid;
    pid = GetProcessId(process_handle);
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+    /* Get LocalDumps directory path. */
+    HKEY registry_key;
+    DWORD ret = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                  L"SOFTWARE\\Microsoft\\Windows\\Windows Error Reporting\\LocalDumps",
```
style nit: align trailing args with the first one (or 4 space indent, if absolutely necessary to fit the line width)

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+                          &dump_folder_len);
+      if (ret != ERROR_SUCCESS) {
+        /* Default value for `dump_folder` is `%LOCALAPPDATA%\CrashDumps`. */
+        WCHAR * localappdata;
```
style nit
```suggestion
        WCHAR* localappdata;
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+                          RRF_RT_ANY,
+                          &key_type,
+                          (PVOID) dump_folder,
+                          &dump_folder_len);
```
```suggestion
                         NULL,
                         L"DumpFolder",
                         RRF_RT_ANY,
                         &key_type,
                         (PVOID) dump_folder,
                         &dump_folder_len);
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+      CreateDirectoryW(dump_folder, NULL);
+
+      /* Construct dump filename. */
+      WCHAR dump_name[MAX_PATH];
```
style nit: move decls to top

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+      WCHAR dump_name[MAX_PATH];
+      _snwprintf_s(dump_name, sizeof(dump_name), _TRUNCATE, L"%ls\\%ls.%d.dmp", dump_folder, basename, pid);
+
+      HANDLE hDumpFile = CreateFileW(dump_name, GENERIC_WRITE, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
```
style nit: move decls to top

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+        SetFileInformationByHandle(hDumpFile, FileDispositionInfo, &DeleteOnClose, sizeof(DeleteOnClose));
+
+        /* Tell wine to dump ELF modules as well. */
+        DWORD sym_options = SymGetOptions();
```
style nit: move decls to top

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+   * for more detail.  Note that if the dump folder does not exist, we attempt
+   * to create it, to match behavior with WER itself.
+   * [0]: https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps
+   */
```
style nit: no blank lines in comments in this file (IIRC)
```suggestion
  /* Create a dump file for the targeted process, if the registry key
   * `HKLM:Software\Microsoft\Windows\Windows Error Reporting\LocalDumps`
   * exists.  The location of the dumps can be influenced by the `DumpFolder`
   * sub-key, which has a default value of `%LOCALAPPDATA%\CrashDumps`, see [0]
   * for more detail.  Note that if the dump folder does not exist, we attempt
   * to create it, to match behavior with WER itself.
   * [0]: https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps */
```

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+                              * to create, as not all clients may wish to create such a large dump.
+                              */
+                              MiniDumpWithFullMemory | MiniDumpIgnoreInaccessibleMemory | MiniDumpWithAvxXStateContext,
+                              NULL, NULL, NULL) == TRUE) {
```
style nit: each argument should usually get its own line. Might be worth moving this into a temporary variable also, but not strictly necessary (nor is `== TRUE` strictly necessary either, and a bit like adding `=== true` in julia to a boolean comparison)

vtjnash commented on src/win/process.c:
```diff
@@ -1250,7 +1253,87 @@ static int uv__kill(HANDLE process_handle, int signum) {
+                              MiniDumpWithFullMemory | MiniDumpIgnoreInaccessibleMemory | MiniDumpWithAvxXStateContext,
+                              NULL, NULL, NULL) == TRUE) {
+          /* Don't delete the file on close if we successfully wrote it out. */
+          FILE_DISPOSITION_INFO DeleteOnClose = { FALSE };
```
style nit: move decls to top of scope (or preferable, top of the function)

vtjnash reviewed (commented):
basically just commenting on a lot of style nits for libuv-style, but otherwise SGTM

staticfloat commented:
Upstream patch submitted here: https://github.com/libuv/libuv/pull/3840

staticfloat commented:
@vtjnash there have been a few times recently where this kind of support would have been helpful.  What do we think; should I bump the upstream PR?

vtjnash commented:
Yeah, go ahead and bump it and I will try to find time (hah)

staticfloat commented:
Now that upstream has been merged, what do we do here?
